core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import io.trino.operator.exchange.PageReference.PageReleasedListener;
+import io.trino.spi.Page;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scale up local writers based on throughput and physical written bytes.
+ * Input pages are distributed across different writers in a round-robin fashion.
+ */
+public class ScaleWriterExchanger
+        implements LocalExchanger
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final List<Consumer<PageReference>> buffers;
+    private final LocalExchangeMemoryManager memoryManager;
+    private final long bufferSizeMinScalingThreshold;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+    private final PageReleasedListener onPageReleased;
+
+    // Start with single writer and increase the writer count based on
+    // physical written bytes and buffer utilization.
+    private int writerCount = 1;
```

Comment:
Doesn't it need to be `synchronized`, e.g: `io.trino.operator.exchange.PartitioningExchanger#partitionAssignments` is (cc @pettyjamesm do you recall why it's synchronized)?

Reply:
For PartitioningExchanger, the “IntArrayList[] partitionAssignments” field is reused between invocations and needs to prevent separate drivers from partitioning pages concurrently in the shared exchanger instance.

Reply:
> field is reused between invocations and needs to prevent separate drivers from partitioning pages concurrently in the shared exchanger instance.

@pettyjamesm I don't think exchanged is shared between drivers, is it?

Reply:
Looking through the code, it seems like they’re not shared- so maybe that changed at some point along the way, I’m not sure. When I refactored partitioning exchanger it was already synchronized so I might have just assumed that they might be shared under some circumstances and kept it synchronized.